execution/block-partitioner/src/v2/partition_to_matrix.rs

```diff
@@ -0,0 +1,189 @@
+        let _timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["partition_to_matrix"])
+            .start_timer();
+
+        let mut remaining_txns = mem::take(&mut state.pre_partitioned);
+        assert_eq!(state.num_executor_shards, remaining_txns.len());
+
+        let mut num_remaining_txns = usize::MAX;
+        for round_id in 0..(state.num_rounds_limit - 1) {
+            let (accepted, discarded) = Self::discarding_round(state, round_id, remaining_txns);
+            state.finalized_txn_matrix.push(accepted);
+            remaining_txns = discarded;
+            num_remaining_txns = remaining_txns.iter().map(|ts| ts.len()).sum();
+
+            if num_remaining_txns < state.avoid_pct as usize * state.num_txns() / 100 {
+                break;
+            }
+        }
+
+        let _timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["last_round"])
+            .start_timer();
+
+        if state.merge_discarded {
+            trace!("Merging txns after discarding stopped.");
+            let last_round_txns: Vec<PreParedTxnIdx> =
+                remaining_txns.into_iter().flatten().collect();
+            remaining_txns = vec![vec![]; state.num_executor_shards];
+            remaining_txns[state.num_executor_shards - 1] = last_round_txns;
+        }
+
+        let last_round_id = state.finalized_txn_matrix.len();
+        state.thread_pool.install(|| {
+            (0..state.num_executor_shards)
+                .into_par_iter()
+                .for_each(|shard_id| {
+                    remaining_txns[shard_id]
+                        .par_iter()
+                        .for_each(|&ori_txn_idx| {
+                            state.update_trackers_on_accepting(
+                                ori_txn_idx,
+                                last_round_id,
+                                shard_id,
+                            );
+                        });
+                });
+        });
+        state.finalized_txn_matrix.push(remaining_txns);
+    }
+
+    /// Given some pre-partitioned txns, pull some off from each shard to avoid cross-shard conflict.
+    /// The pulled off txns become the pre-partitioned txns for the next round.
+    pub(crate) fn discarding_round(
+        state: &mut PartitionState,
+        round_id: RoundId,
+        remaining_txns: Vec<Vec<PreParedTxnIdx>>,
+    ) -> (Vec<Vec<PreParedTxnIdx>>, Vec<Vec<PreParedTxnIdx>>) {
+        let _timer = MISC_TIMERS_SECONDS
+            .with_label_values(&[format!("round_{round_id}").as_str()])
+            .start_timer();
+
+        let num_shards = remaining_txns.len();
+        let mut discarded: Vec<RwLock<Vec<PreParedTxnIdx>>> = Vec::with_capacity(num_shards);
+        let mut tentatively_accepted: Vec<RwLock<Vec<PreParedTxnIdx>>> =
+            Vec::with_capacity(num_shards);
+        let mut finally_accepted: Vec<RwLock<Vec<PreParedTxnIdx>>> = Vec::with_capacity(num_shards);
```

Comment:
Add some comment explaining the difference between `tentatively_accepted` and `finally_accepted`